### PR TITLE
OCEANWATER-346: Use a different topic name for dynamic terrain model and visual plugins

### PR DESCRIPTION
**Pull request by Samahu**
## Summary of changes:
* Collision between the terrain and the end effectors is now enabled
* Topics corresponding to /ow_dynamic_terrain/modify_terrain_* are left intact. Plugin-specific topics {visual and collision} are added to the *DynamicTerrainVisual* and *DynamicTerrainModel* plugins
* Python scripts `modify_terrain_scoop_pub.py` and `modify_terrain_grinder_pub.py` corresponding to the scoop and grinder has been updated to use the Plugin-specific topics with a slightly different parameters for visual vs collision.
 - The `modify_terrain_*/collision` message is merely a scaled up version of the `modify_terrain_*/visual` message  
* README file updated with the new changes

## Verify
Launch the simulation against the `europa_terminator_workspace`:
```bash
roslaunch ow europa_terminator_workspace.launch
```

Use the rqt service caller to perform any arm planning operation that may interact with the terrain. e.g.
- /planning/arm/dig_circular
- /planning/arm/dig_linear
- /planning/arm/grind
Followed by a service call to `/planning/arm/publish_trajectory` with use_latest=True

**Observe the arm modifies the terrain and the arm is not colliding with the terrain.**

## Experiments 
To see the actual size of the terrain set the `modify_terrain_*/visual` message scale value in the  `modify_terrain_scoop_pub.py` and `modify_terrain_grinder_pub.py` to the same scale value of the `modify_terrain_*/collision` message.

### CODE REVIEW

**Comment by damianacatanoso** [comment]
Get "unstable" behavior similar to guarded_move: https://youtu.be/STjibPuHp4Y

**Comment by damianacatanoso** [comment]
Here is another instance of the same problem when publishing DigLinear: https://youtu.be/nJH5f5cYCy4

**Comment by kmdalal** [comment]
I see the same thing in Rviz.

﻿On 8/18/20, 3:31 PM, "damianacatanoso" <notifications@github.com> wrote:

    Get "unstable" behavior similar to guarded_move: https://youtu.be/STjibPuHp4Y

**Comment by damianacatanoso** [comment]
> I see the same thing in Rviz. ﻿On 8/18/20, 3:31 PM, "damianacatanoso" <notifications@github.com> wrote: Get "unstable" behavior similar to guarded_move: https://youtu.be/STjibPuHp4Y

Mike, do you run Gazebo at all?

**Comment by kmdalal** [comment]
Just the server.  The client (GUI) never ran in any VM for me.  I may try to adjust the VM resources again, but I am planning to get a native Linux machine.

﻿On 8/18/20, 3:40 PM, "damianacatanoso" <notifications@github.com> wrote:


    I see the same thing in Rviz. On 8/18/20, 3:31 PM, "damianacatanoso" notifications@github.com wrote: Get "unstable" behavior similar to guarded_move: https://youtu.be/STjibPuHp4Y


    Mike, do you run Gazebo at all?

**Comment by damianacatanoso** [comment]
> Just the server. The client (GUI) never ran in any VM for me. I may try to adjust the VM resources again, but I am planning to get a native Linux machine. ﻿On 8/18/20, 3:40 PM, "damianacatanoso" <notifications@github.com> wrote: I see the same thing in Rviz. On 8/18/20, 3:31 PM, "damianacatanoso" notifications@github.com wrote: Get "unstable" behavior similar to guarded_move: https://youtu.be/STjibPuHp4Y Mike, do you run Gazebo at all?

Oh I see, you are running the server. It makes sense we see the same behaviour!

**Comment by damianacatanoso** [comment]
@Samahu I just pulled ow_simulator. Should I pull ow_europa or others as well?

**Comment by kmdalal** [comment]
Yeah, it wasn't obvious to me what repo this branch is for.  I just guessed only ow_simulator.

**Comment by damianacatanoso** [comment]
> Yeah, it wasn't obvious to me what repo this branch is on. I just guessed only ow_simulator.

If you place your cursor on "master" or "OCEANWATER-346_Use_a_diffe..[name of branch]" it explicitely says the repo belongs to ow_simulator. I asked just to be sure!

File: ow_dynamic_terrain/src/DynamicTerrainBase.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+protected:
+  std::unique_ptr<ros::NodeHandle> m_node_handle;
+
+protected:
+  ros::CallbackQueue m_callback_queue;
+
+protected:
+  std::vector<ros::Subscriber> m_subscribers;
+};
```
[the remark this answers is not in the document]
Comment (Samahu) [reply ow_dynamic_terrain/src/DynamicTerrainBase.h:63]: It is something I picked from Gazebo code (https://github.com/osrf/gazebo/blob/gazebo9/gazebo/rendering/Heightmap.hh). I don't always use this style but since this was a Gazebo plugin I went with it. No strong preference, I like it partially because it it aligns with C# and other languages and I don't need to worry much about grouping fields that have the same access modifier.

**Comment by Samahu** [comment]
> Get "unstable" behavior similar to guarded_move: https://youtu.be/STjibPuHp4Y

@damianacatanoso I didn't see this while testing but recently I was only testing with the grinder. I'll try and let you know.

>Note: The changes are only made to ow_simulator

**Comment by Samahu** [comment]
@damianacatanoso @mikedalal  I did verify that the same problems shows up on my machine. I was not sure at what point this problem started to take place. The problem wasn't occurring when I was testing against the scoop otherwise I would have noticed it. After trying few things I found out that the problem actually started when I enabled collision between the terrain and the grinder

I don't know exactly why as the grinder is not effectively touching the terrain in the case of a circle  dig.

**Comment by mogumbo** [comment]
I have only had time to try dig_circular. I see the same arm jumping motion that Damiana recorded in her video. I've been working on the move_guarded problem in OCEANWATER-399. Using the quick solver instead of world solver in the .world file helps a lot. Unfortunately, in this case, merging my OCEANWATER-399 branch in ow_simulator did not help any more. It has smaller arm torque limits that added stability in other test cases, but it did not help this situation. Anyway, I see this as a problem unrelated to this ticket.

Super cool digging, by the way!

File: ow_dynamic_terrain/src/DynamicTerrainBase.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mogumbo) [reply ow_dynamic_terrain/src/DynamicTerrainBase.h:62]: I prefer removing the redundant lines. Also, I find it useful to group public, private, and protected methods together as much as possible. It reduces the code size because you need fewer access qualifiers , and it's easier to learn to use a class when you can see all its public methods together in the same place.

File: ow_dynamic_terrain/src/DynamicTerrainBase.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  if (!ros::isInitialized())
+  {
+    gzerr << m_plugin_name + ": ROS not initilized! The plugin won't load" << endl;
+    return;
+  }
+
+  m_node_handle = make_unique<ros::NodeHandle>(m_plugin_name);
+  m_node_handle->setCallbackQueue(&m_callback_queue);
+
+  auto on_modify_terrain_circle = [this](const modify_terrain_circle::ConstPtr& msg) {
```
Comment (mogumbo) [inline ow_dynamic_terrain/src/DynamicTerrainBase.cpp:18]: This is a good use for auto.

File: ow_dynamic_terrain/src/DynamicTerrainBase.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  });
+
+  gzlog << m_plugin_name << ": successfully loaded!" << endl;
+}
+
+template <typename T>
+void DynamicTerrainBase::subscribe(const std::string& topic,
+                                   const boost::function<void(const boost::shared_ptr<T const>&)>& callback)
+{
+  auto topic_fqn = "/" + m_package_name + "/" + topic;
```
Comment (mogumbo) [inline ow_dynamic_terrain/src/DynamicTerrainBase.cpp:48]: Please use simple type names instead of auto.

File: ow_dynamic_terrain/src/DynamicTerrainBase.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+gazebo::rendering::Heightmap* DynamicTerrainBase::getHeightmap(gazebo::rendering::ScenePtr scene)
+{
+  if (!scene)
+  {
+    gzerr << m_plugin_name << ": Couldn't acquire scene!" << endl;
+    return nullptr;
+  }
+
+  auto heightmap = scene->GetHeightmap();
```
Comment (mogumbo) [inline ow_dynamic_terrain/src/DynamicTerrainBase.cpp:60]: Here too. This is a simple type name.

File: ow_dynamic_terrain/scripts/modify_terrain_scoop_pub.py [hunk 5]
```diff
@@ -15,23 +15,24 @@
...
         "/gazebo/link_states", LinkStates, self.handle_link_states)
 
-  def compose_modify_terrain_ellipse_message(self, position, orientation):
+  def compose_modify_terrain_ellipse_message(self, position, orientation, scale=1.0):
     return modify_terrain_ellipse(position=Point(position.x, position.y, position.z),
                                   orientation=orientation,
-                                  outer_radius_a=0.002, outer_radius_b=0.005,
-                                  inner_radius_a=0.001, inner_radius_b=0.001,
-                                  weight=-0.025, merge_method="min")
+                                  outer_radius_a=0.002*scale, outer_radius_b=0.005*scale,
```
Comment (mogumbo) [inline ow_dynamic_terrain/scripts/modify_terrain_scoop_pub.py:33]: Where do the numbers 0.002, 0.005, and 0.001 come from? Please add a comment to explain the logic here.

File: ow_dynamic_terrain/scripts/modify_terrain_grinder_pub.py [hunk 6]
```diff
@@ -21,15 +21,17 @@ class ModifyTerrainGrinder:
...
+        'ow_dynamic_terrain/modify_terrain_circle/collision', modify_terrain_circle, queue_size=1)
     self.states_sub = rospy.Subscriber(
         "/gazebo/link_states", LinkStates, self.handle_link_states)
 
-  def compose_modify_terrain_circle_message(self, position):
+  def compose_modify_terrain_circle_message(self, position, scale=1.0):
     return modify_terrain_circle(position=Point(position.x, position.y, position.z),
-                                 outer_radius=0.008, inner_radius=0.001,
-                                 weight=-0.2, merge_method="min")
+                                 outer_radius=0.008*scale, inner_radius=0.001*scale,
```
Comment (mogumbo) [inline ow_dynamic_terrain/scripts/modify_terrain_grinder_pub.py:33]: Where do the numbers 0.008 and 0.001 come from? Please add a comment to explain the logic here. If there is to be a default size for a circle, I would expect it to be a unit circle.

File: ow_dynamic_terrain/include/memory_ext.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+// TODO: consider moving the file to a common place
+
+#include <memory>
+
+#if __cplusplus < 201402L  // if less than C++14 is detected add the extension
+
+namespace std
+{
+template <typename T, typename... Args>
```
Comment (mogumbo) [inline ow_dynamic_terrain/include/memory_ext.h:12]: Please comment on what these parts do.

File: ow_lander/urdf/lander.xacro [hunk 8]
```diff
@@ -543,15 +543,6 @@
       </plugin>
     </visual>
     <selfCollide>1</selfCollide>
-    <!-- Turn off collisions with the terrain so scoop can go underground. For
-         this to work, the terrain must use a different collide_bitmask. -->
-    <collision>
-      <surface>
```
Comment (mogumbo) [inline ow_lander/urdf/lander.xacro:549]: Collisions are currently disabled for the hand as well. We should probably also remove that code block, unless there is some reason to keep it that I don't know about.

**Comment by damianacatanoso** [comment]
FYI, I was trying drilling/digging in other directions (use_default=False). I used the grinder to dig a first hole. Then I used the grinder a second time to extend the previous hole. Even if I used only the grinder, I observed the same arm bouncing behavior that we observed at the end of dig linear/dig circular.

File: ow_dynamic_terrain/include/memory_ext.h (hunk 7 above)
Comment (Samahu) [reply ow_dynamic_terrain/include/memory_ext.h:21]: addressed

File: ow_dynamic_terrain/scripts/modify_terrain_grinder_pub.py (hunk 6 above)
Comment (Samahu) [reply ow_dynamic_terrain/scripts/modify_terrain_grinder_pub.py:42]: addressed

File: ow_dynamic_terrain/include/memory_ext.h (hunk 7 above)
Comment (Samahu) [reply ow_dynamic_terrain/include/memory_ext.h:21]: C++11 forgot to add a make_unique function similar to make_shared so I had it added for C++11 only compiles

I think it is beneficial to add this in a common place such that  other modules can make use of it

File: ow_dynamic_terrain/scripts/modify_terrain_scoop_pub.py (hunk 5 above)
Comment (Samahu) [reply ow_dynamic_terrain/scripts/modify_terrain_scoop_pub.py:42]: addressed

File: ow_dynamic_terrain/src/DynamicTerrainBase.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Samahu) [reply ow_dynamic_terrain/src/DynamicTerrainBase.h:47]: Gazebo applies this consistently through their code and it is not impeding me personally. I have also seen this applied in other frameworks and API. In any case, since there are two votes against one, I have the properties grouped.

File: ow_lander/urdf/lander.xacro (hunk 8 above)
Comment (Samahu) [reply ow_lander/urdf/lander.xacro:549]: I did try this before actually submitting this PR and it did actually cause slowness to happen

File: ow_dynamic_terrain/src/DynamicTerrainBase.cpp (hunk 4 above)
Comment (Samahu) [reply ow_dynamic_terrain/src/DynamicTerrainBase.cpp:64]: The simple type rule is really very hard to follow. I don't know which types are simple and which types are not. for example initially we pointed at integers and floats as simple types. Now we considering std::string a simple type and gazebo::rendering::Heightmap* as a simple type.

Comment (mogumbo) [reply ow_dynamic_terrain/src/DynamicTerrainBase.cpp:64]: Yes, those are simple types, although I didn't call them out explicitly in the coding standard. It's easy to type either of those and easy to understand what they are. Using auto here makes the code smaller but harder to understand. I'm not opposed to using namespace gazebo::rendering to shorten things a bit, but some people might disagree with me on that one.

File: ow_lander/urdf/lander.xacro (hunk 8 above)
Comment (mogumbo) [reply ow_lander/urdf/lander.xacro:549]: Weird. If we're going to keep it, it would be a good idea to amend that comment to explain that is is a performance booster. Later on we should think about disabling collisions between other arm parts and the terrain; maybe we can get even more performance boost.

File: ow_dynamic_terrain/src/DynamicTerrainBase.cpp (hunk 2 above)
Comment (Samahu) [reply ow_dynamic_terrain/src/DynamicTerrainBase.cpp:22]: I don't agree to this narrative. It is almost always good to use **auto**. In any case, I have amended the PR to declare the type of each variable as I don't wish to delay the PR any further.

File: ow_dynamic_terrain/src/DynamicTerrainBase.cpp (hunk 3 above)
Comment (Samahu) [reply ow_dynamic_terrain/src/DynamicTerrainBase.cpp]: addressed

File: ow_dynamic_terrain/src/DynamicTerrainBase.cpp (hunk 4 above)
Comment (Samahu) [reply ow_dynamic_terrain/src/DynamicTerrainBase.cpp]: Again, I don't agree that the use of **auto** makes the code harder to read and understand (best counter example is python). In any case, this has been addressed.

File: ow_lander/urdf/lander.xacro (hunk 8 above)
Comment (Samahu) [reply ow_lander/urdf/lander.xacro:549]: addressed
